### ISSUE TITLE
redundant_closure_for_method_calls suggestion introduces type error

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#9900]
### Summary

.

### Lint Name

redundant_closure_for_method_calls

### Reproducer

I tried this code:

```rust
// run-pass
#![allow(dead_code)]
// pretty-expanded FIXME #23616

trait Typer<'tcx> {
    fn method(&self, data: &'tcx isize) -> &'tcx isize { data }
    fn dummy(&self) { }
}

fn g<F>(_: F) where F: FnOnce(&dyn Typer) {}

fn h() {
    g(|typer| typer.dummy())
}

fn main() { }
```

I saw this happen:
`cargo clippy --fix -- -Aclippy::all -Wclippy::redundant_closure_for_method_calls`
```
    Checking clpy v0.1.0 (/tmp/clpy)
warning: failed to automatically apply fixes suggested by rustc to crate `clpy`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0308]: mismatched types
  --> src/main.rs:13:5
   |
13 |     g(Typer::dummy)
   |     ^ one type is more general than the other
   |
   = note: expected trait `for<'a, 'b> std::ops::FnOnce<(&'a (dyn Typer<'b> + 'a),)>`
              found trait `for<'a> std::ops::FnOnce<(&'a dyn Typer<'_>,)>`
note: the lifetime requirement is introduced here
  --> src/main.rs:10:24
   |
10 | fn g<F>(_: F) where F: FnOnce(&dyn Typer) {}
   |                        ^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
Original diagnostics will follow.

warning: redundant closure
  --> src/main.rs:13:7
   |
13 |     g(|typer| typer.dummy())
   |       ^^^^^^^^^^^^^^^^^^^^^ help: replace the closure with the method itself: `Typer::dummy`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure_for_method_calls
   = note: requested on the command line with `-W clippy::redundant-closure-for-method-calls`

warning: `clpy` (bin "clpy") generated 1 warning (1 duplicate)
warning: `clpy` (bin "clpy" test) generated 1 warning (run `cargo fix --bin "clpy" --tests` to apply 1 suggestion)
    Finished dev [unoptimized + debuginfo] target(s) in 0.44s
```

I expected to see this happen:

### Version

```text
rustc 1.67.0-nightly (c5d82ed7a 2022-11-19)
binary: rustc
commit-hash: c5d82ed7a4ad94a538bb87e5016e7d5ce0bd434b
commit-date: 2022-11-19
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```

**Comment by profetia** [comment]
This is no longer reproducible in the latest nightly

**Comment by ada4a** [comment]
I wasn't able to (quickly) find out which PR fixed this, but I confirmed [on Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=18eaed3d41a760076a85aa2d5ffc1d73) that this is not reproducible.

@rustbot label S-fixed

**Comment by jpmckinney** [comment]
As another example, I can confirm that I previously had:

```rust
    // https://github.com/rust-lang/rust-clippy/issues/9900
    #[allow(clippy::redundant_closure_for_method_calls)]
    fn flush(&mut self) -> Result<()> {
        self.apply(|w| w.flush())
    }
```

And now I no longer need the `allow`.
